src/function/table/call/hnsw/query_hnsw_index.cpp

```diff
@@ -35,39 +36,97 @@ static std::unique_ptr<TableFuncBindData> bindFunc(main::ClientContext* context,
-    storage::HNSWIndexUtils::validateQueryVector(queryVal.getDataType(), dimension);
-
-    std::vector<float> queryVector;
-    KU_ASSERT(queryVal.getChildrenSize() > 0);
-    queryVector.resize(queryVal.getChildrenSize());
-    for (auto i = 0u; i < queryVector.size(); i++) {
-        queryVector[i] = common::NestedVal::getChildVal(&queryVal, i)->getValue<float>();
-    }
+    KU_ASSERT(
+        nodeTableEntry.getProperty(indexEntry->getIndexColumnName()).getType().getLogicalTypeID() ==
+        common::LogicalTypeID::ARRAY);
 
     auto outputNode = input->binder->createQueryNode("nn", {&nodeTableEntry});
     input->binder->addToScope(outputNode->toString(), outputNode);
     binder::expression_vector columns;
     columns.push_back(outputNode->getInternalID());
     columns.push_back(input->binder->createVariable("_distance", common::LogicalType::DOUBLE()));
-    auto boundInput = BoundQueryHNSWIndexInput{&nodeTableEntry, indexEntry, std::move(queryVector),
-        static_cast<uint64_t>(k)};
+    auto boundInput = BoundQueryHNSWIndexInput{&nodeTableEntry, indexEntry,
+        std::move(queryExpression), static_cast<uint64_t>(k)};
     auto config = storage::QueryHNSWConfig{input->optionalParams};
     return std::make_unique<QueryHNSWIndexBindData>(context, std::move(columns), boundInput, config,
         outputNode);
 }
 
+static common::PhysicalTypeID getChildQueryValType(const common::Value& value) {
+    common::PhysicalTypeID childValType;
+    switch (value.getDataType().getPhysicalType()) {
+    case common::PhysicalTypeID::ARRAY: {
+        childValType = value.getDataType()
+                           .getExtraTypeInfo()
+                           ->constPtrCast<common::ArrayTypeInfo>()
+                           ->getChildType()
+                           .getPhysicalType();
+    } break;
+    case common::PhysicalTypeID::LIST: {
+        childValType = value.getDataType()
+                           .getExtraTypeInfo()
+                           ->constPtrCast<common::ListTypeInfo>()
+                           ->getChildType()
+                           .getPhysicalType();
+    } break;
+    default: {
+        throw common::RuntimeException(
+            common::stringFormat("Unsupported data type {} as a query vector in {}",
+                value.getDataType().toString(), QueryHNSWIndexFunction::name));
+    }
+    }
+    return childValType;
+}
+
+static std::vector<float> getQueryVector(const binder::Expression& queryExpression,
+    uint64_t dimension) {
+    auto value = common::Value::createDefaultValue(queryExpression.dataType);
+    switch (queryExpression.expressionType) {
+    case common::ExpressionType::LITERAL: {
+        value = queryExpression.constCast<binder::LiteralExpression>().getValue();
+    } break;
+    case common::ExpressionType::PARAMETER: {
+        value = queryExpression.constCast<binder::ParameterExpression>().getValue();
+    } break;
+    default: {
+        KU_UNREACHABLE;
```

Comment:
What will happen if expression is not of these two types in release mode? I think it's better to throw here.